version: '3.8'

services:
  backend:
    image: node:18.20-alpine3.20
    build: 
      context: ./backend
    ports:
      - '3001:3001'
    volumes:
      - ./backend:/app
      # - /app/node_modules
    depends_on:
      db:
        condition: service_healthy
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=beer_reviews
      - PGHOST=db
      - PGPORT=5432
      - PGUSER=postgres
      - PGPASSWORD=postgres
      - PGDATABASE=beer_reviews
    networks:
      - app-network

  frontend:
    image: node:18.20-alpine3.20
    build:
      context: ./frontend
    ports:
      - '3000:3000'
    networks:
      - app-network
    volumes:
      - ./frontend:/app
      - /app/node_modules

  db:
    image: postgres:16-alpine3.20
    restart: always
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: beer_reviews
    volumes:
      - ./db/init:/docker-entrypoint-initdb.d
      - db-data:/var/lib/postgresql/data
    networks:
      - app-network
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres -d beer_reviews"]
      interval: 5s
      timeout: 5s
      retries: 5

networks:
  app-network:
    driver: bridge

volumes:
  db-data:
